### PR TITLE
[C++ API] Turn BatchNorm into BatchNorm1d,2d,3d

### PR DESCRIPTION
**Pull request by goldsborough**
@jamespinkerton mentioned that he found it confusing that in the C++ API there is only `BatchNorm`, while in PyTorch we have `BatchNorm{1,2,3}d`. This PR changes this.

It's the same module/function call in the end, but I took the opportunity to add another dimensionality check to every of the BatchNormXd modules, like PyTorch https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L52

@ebetica @apaszke @ezyang

### CODE REVIEW

**Comment by SsnL** [comment]
That we have those `Nd` thing was purely for BC reason. Unless we really really want to match Python and c++ APIs on every bit, I would drop the `Nd` suffix.

**Review by ezyang — APPROVED** [review]
Sure why not

**Comment by goldsborough** [comment]
@SsnL it's useful to match the Python naming so that it's easier for people to rewrite their models from Python into C++.

**Comment by ebetica** [comment]
Meh, I think I agree with @SsnL, this is a place where better documentation is needed, not PyTorch parity.

**Comment by colesbury** [comment]
To chime in here, I agree with @SsnL's original point. I don't think it's worth adding a bunch of identical classes with different names in C++.

**Comment by goldsborough** [comment]
Ok, that's fine with me too then.

**Comment by fmassa** [comment]
+1 on not duplicating batch norm. Also, maybe we should consider deprecating batch norm {1-3}d in favour of batch norm ? Relevant issue https://github.com/pytorch/pytorch/issues/2628

**Comment by goldsborough** [comment]
@fmassa that sounds like a good idea. I agree that it's confusing since it makes it look like the operations do something different in some way (I was purely after making it easier for Python users to migrate here). Keras and the like also only provide one BatchNorm op
